### PR TITLE
Reapply "[LICM] Fold associative binary ops to promote code hoisting (#81608)"

### PR DESCRIPTION
**Pull request by rj-jesus**
This reverts commit https://github.com/llvm/llvm-project/commit/b48819dbcdb48fc737dc22304ac343e4fdbae9ff.

This reapplies a more strict version of https://github.com/llvm/llvm-project/commit/f2ccf80136a01ca69f766becafb329db6c54c0c8.

Perform the transformation

  "(LV op C1) op C2" ==> "LV op (C1 op C2)"

where op is an associative binary op, LV is a loop variant, and C1 and C2 are loop invariants, and hoist (C1 op C2) into the preheader.

For now this fold is restricted to ADDs.

### CODE REVIEW

File: llvm/lib/Transforms/Scalar/LICM.cpp [hunk 1]
```diff
@@ -2779,6 +2781,72 @@ static bool hoistMulAddAssociation(Instruction &I, Loop &L,
...
+        BO0->hasNoUnsignedWrap()) {
+      Inv->setHasNoUnsignedWrap(true);
+      NewBO->setHasNoUnsignedWrap(true);
+    }
+
+    BO->replaceAllUsesWith(NewBO);
+    eraseInstruction(*BO, SafetyInfo, MSSAU);
+
+    // Note: (LV op C1) might not be erased if it has more uses than the one we
+    //       just replaced.
```
Comment (nikic) [inline llvm/lib/Transforms/Scalar/LICM.cpp:2840]:
```suggestion
    // (LV op C1) might not be erased if it has more uses than the one we
    // just replaced.
```

File: llvm/lib/Transforms/Scalar/LICM.cpp [hunk 2]
```diff
@@ -2779,6 +2781,72 @@ static bool hoistMulAddAssociation(Instruction &I, Loop &L,
...
+
+  // Only fold ADDs for now.
+  Instruction::BinaryOps Opcode = BO->getOpcode();
+  if (Opcode != Instruction::Add)
+    return false;
+
+  BinaryOperator *BO0 = dyn_cast<BinaryOperator>(BO->getOperand(0));
+
+  // Transform: "(LV op C1) op C2" ==> "LV op (C1 op C2)"
+  if (BO0 && BO0->getOpcode() == Opcode && BO0->isAssociative()) {
```
Comment (nikic) [inline llvm/lib/Transforms/Scalar/LICM.cpp:2811]: Early exit.

File: llvm/test/Transforms/LICM/hoist-binop.ll [hunk 3]
```diff
@@ -0,0 +1,229 @@
...
+
+loop:
+  %index = phi i64 [ 0, %entry ], [ %index.next, %loop ]
+  %step.add = sub i64 %index, %c
+  call void @use(i64 %step.add)
+  %index.next = sub i64 %step.add, %c
+  br label %loop
+}
+
+; Don't fold floating-point ops, even if they are associative.
```
Comment (nikic) [inline llvm/test/Transforms/LICM/hoist-binop.ll:172]: Maybe clarify here that the transform is valid, we just don't do it yet.

**Review by nikic — COMMENTED** [review]
Looks reasonable, some nits.

File: llvm/lib/Transforms/Scalar/LICM.cpp [hunk 4]
```diff
@@ -2779,6 +2781,72 @@ static bool hoistMulAddAssociation(Instruction &I, Loop &L,
...
+    auto *Preheader = L.getLoopPreheader();
+    assert(Preheader && "Loop is not in simplify form?");
+
+    auto *Inv = BinaryOperator::Create(Opcode, C1, C2, "invariant.op",
+                                       Preheader->getTerminator());
+    auto *NewBO = BinaryOperator::Create(Opcode, LV, Inv,
+                                         BO->getName() + ".reass", BO);
+
+    // Copy NUW for ADDs if both instructions have it.
+    // https://alive2.llvm.org/ce/z/K9W3rk
```
[the remark this answers is not in the document]
Comment (rj-jesus) [reply llvm/lib/Transforms/Scalar/LICM.cpp]: Thanks, I've removed the link.

**Comment by rj-jesus** [comment]
@nikic Thanks for the comments, I've made the suggested changes.

**Review by nikic — APPROVED** [review]
LGTM

**Comment by rj-jesus** [comment]
Thanks!

**Comment by dtcxzyw** [comment]
Hi @rj-jesus @nikic

My CI reported that this patch caused significant performance regressions on `SingleSource/Benchmarks/Stanford/Queens` (+50% instructions) and `SingleSource/Benchmarks/Misc/himenobmtxpa` (+20% instructions). Can you have a look?
See also https://github.com/dtcxzyw/llvm-ci/issues/1176.

**Comment by rj-jesus** [comment]
Hi @dtcxzyw 

Thanks for reporting this. I can't seem to reproduce the performance regression on `SingleSource/Benchmarks/Stanford/Queens` but do see a 5% increase in `SingleSource/Benchmarks/Misc/himenobmtxpa` and some increases in code size.

I think this can be addressed by bailing out early if `BO0->getNumUses() > 2`. This avoids cases where the intermediate op we are trying to avoid has many uses, and so performing the fold and hoist will likely not be beneficial or lead to a significant increase in code size. This seems to be what's happening with `himenobmtxpa`.

File: llvm/lib/Transforms/Scalar/LICM.cpp [hunk 5]
```diff
@@ -2779,6 +2781,68 @@ static bool hoistMulAddAssociation(Instruction &I, Loop &L,
...
+      BinaryOperator::Create(Opcode, LV, Inv, BO->getName() + ".reass", BO);
+
+  // Copy NUW for ADDs if both instructions have it.
+  if (Opcode == Instruction::Add && BO->hasNoUnsignedWrap() &&
+      BO0->hasNoUnsignedWrap()) {
+    Inv->setHasNoUnsignedWrap(true);
+    NewBO->setHasNoUnsignedWrap(true);
+  }
+
+  BO->replaceAllUsesWith(NewBO);
```
Comment (svs-quic) [inline llvm/lib/Transforms/Scalar/LICM.cpp:2835]: Hi, do we need to preserve the debug information of the original binary operator by copying it over to the new one?

Something like:
`NewBO->setDebugLoc(BO->getDebugLoc());`

Comment (rj-jesus) [reply llvm/lib/Transforms/Scalar/LICM.cpp:2835]: Hi, I think you're right, thanks. I'll have a look.

**Pull request merged**